### PR TITLE
Allow to build the documentation in a release

### PR DESCRIPTION
**Pull request by bo0ts**
This PR adds the ability to build the documentation in the release layout. To do that variables for generic (with respect to the layout) references to example/demo/include directories are added and a filelist.txt file for each package is generated during release creating.

The documentation testsuite of a release reports warnings in packages that have examples in dont_submit, but the examples are shown in the local build. This is a genuine bug and should be addressed separately.

Commit 9642702 and def7234 are almost pure formatting changes which increase the readability and I consider them worth it.

I tried to do as much visual confirmation of the result as possible and checked the testsuite.

### CODE REVIEW

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-69

**Comment by sloriot** [comment]
Using the former way to build the doc, the testsuite looks good.

On my machine, I compared the list of files generated with your branch and 022c7b41 and they are the same (up to some `dir_#sh` I never saw before (btw any idea if these files are useful? if not how to remove them?)).

**Comment by sloriot** [comment]
@bo0ts: How should we run the testsuite from a release?

**Comment by bo0ts** [comment]
> How should we run the testsuite from a release?

This should still work as before. You need to set `CGAL_DOC_CREATE_LOGS` to `TRUE` and have python installed. Then the `doc_test` target should be available.

>  (up to some dir_#sh I never saw before (btw any idea if these files are useful? if not how to remove them?))

I haven't noticed them before. Will investigate.

**Comment by sloriot** [comment]
@bo0ts One last thing: you haven't talk about what is changing for developer of a package. What INPUT should be now?

**Comment by bo0ts** [comment]
> One last thing: you haven't talk about what is changing for developer of a package. What INPUT should be now?

Nothing. It should not be necessary to specify input anymore, unless you want to mix out-of-source documentation with some in-source documentation or you only want to include certain files in the `INPUT` (e.g. BGL package).

As a general rule: 
- `doc/Package` is always part of INPUT
- if the package does not have a `doc/Package/include`, a list of all files in `include` are part of INPUT

**Comment by sloriot** [comment]
I have a few warnings that are not related to the example when I'm building the doc from `CGAL-4.8-Ic-70.tar.gz`.
For example:
```
/tmp/CGAL-4.8-Ic-70/doc/Surface_mesh_segmentation/PackageDescription.txt:6: Warning: explicit link request to 'CGAL::sdf_values()' could not be resolved
/tmp/CGAL-4.8-Ic-70/doc/Surface_mesh_segmentation/PackageDescription.txt:7: Warning: explicit link request to 'CGAL::sdf_values_postprocessing()' could not be resolved
/tmp/CGAL-4.8-Ic-70/doc/Surface_mesh_segmentation/PackageDescription.txt:8: Warning: explicit link request to 'CGAL::segmentation_from_sdf_values()' could not be resolved
/tmp/CGAL-4.8-Ic-70/doc/Surface_mesh_segmentation/PackageDescription.txt:9: Warning: explicit link request to 'CGAL::segmentation_via_sdf_values()' could not be resolved
```
Also I compared the list of files generated with the branch version and a few files are missing in some packages where the doc is in the header. For example:
````
doc_output/AABB_tree/classCGAL_1_1AABB__face__graph__triangle__primitive.html
doc_output/AABB_tree/classCGAL_1_1AABB__face__graph__triangle__primitive.js
doc_output/AABB_tree/classCGAL_1_1AABB__face__graph__triangle__primitive-members.html
doc_output/AABB_tree/classCGAL_1_1AABB__halfedge__graph__segment__primitive.html
doc_output/AABB_tree/classCGAL_1_1AABB__halfedge__graph__segment__primitive.js
doc_output/AABB_tree/classCGAL_1_1AABB__halfedge__graph__segment__primitive-members.html
doc_output/AABB_tree/classCGAL_1_1AABB__polyhedron__segment__primitive.html
doc_output/AABB_tree/classCGAL_1_1AABB__polyhedron__segment__primitive.js
doc_output/AABB_tree/classCGAL_1_1AABB__polyhedron__segment__primitive-members.html
doc_output/AABB_tree/classCGAL_1_1AABB__polyhedron__triangle__primitive.html
doc_output/AABB_tree/classCGAL_1_1AABB__polyhedron__triangle__primitive.js
doc_output/AABB_tree/classCGAL_1_1AABB__polyhedron__triangle__primitive-members.html
doc_output/AABB_tree/classCGAL_1_1AABB__segment__primitive.html
doc_output/AABB_tree/classCGAL_1_1AABB__segment__primitive.js
doc_output/AABB_tree/classCGAL_1_1AABB__segment__primitive-members.html
doc_output/AABB_tree/classCGAL_1_1AABB__traits.html
doc_output/AABB_tree/classCGAL_1_1AABB__traits.js
doc_output/AABB_tree/classCGAL_1_1AABB__traits-members.html
doc_output/AABB_tree/classCGAL_1_1AABB__tree.html
doc_output/AABB_tree/classCGAL_1_1AABB__tree.js
doc_output/AABB_tree/classCGAL_1_1AABB__tree-members.html
doc_output/AABB_tree/classCGAL_1_1AABB__triangle__primitive.html
doc_output/AABB_tree/classCGAL_1_1AABB__triangle__primitive.js
doc_output/AABB_tree/classCGAL_1_1AABB__triangle__primitive-members.html
````

**Comment by bo0ts** [comment]
I could not reproduce the case of missing files on my machine. Some files differed, because the version was not correctly set in the release.

The `dir_23234.html` files are generated through hashing and differ each build as well. 

There should still be some differences between doc build (date generated etc.), but nothing significant.

**Comment by bo0ts** [comment]
Merged to integration in 120737c

**Comment by bo0ts** [comment]
59baa1d should fix the issue of the empty `filelist.txt`.

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-81

**Comment by bo0ts** [comment]
I've confirmed that documentation build from Ic-81 leads to almost identical documentation as build locally.

The only non-generated files that are different is documentation in Point_set_processing_3. The class and concept list (http://doc.cgal.org/latest/Point_set_processing_3/annotated.html) shows a namespace Voronoi_covariance_3. Since that namespace is defined in an internal file it is not listed in documentation build from a release, but it shows up in our current documentation. IMO this is a bug in Point_set_processing and should be fixed there.

**Comment by sloriot** [comment]
Building from the release I see these errors:

    /tmp/CGAL-4.8-Ic-81/doc/Skin_surface_3/examples.txt:6: Warning: included file Skin_surface_3/NGHK_skin_surface_simple.cpp is not found. Check your EXAMPLE_PATH
    /tmp/CGAL-4.8-Ic-81/doc/Skin_surface_3/examples.txt:6: Warning: included file Skin_surface_3/NGHK_skin_surface_subdiv.cpp is not found. Check your EXAMPLE_PATH
    /tmp/CGAL-4.8-Ic-81/doc/Surface_mesher/examples.txt:7: Warning: included file Surface_mesher/polyhedron_remesher.cpp is not found. Check your EXAMPLE_PATH
    /tmp/CGAL-4.8-Ic-81/doc/BGL/examples.txt:19: Warning: included file BGL_surface_mesh/surface_mesh_dual.cpp is not found. Check your EXAMPLE_PATH
    /tmp/CGAL-4.8-Ic-81/doc/Mesh_3/examples.txt:17: Warning: included file Mesh_3/mesh_polyhedral_edge_tolerance_region.cpp is not found. Check your EXAMPLE_PATH
    /tmp/CGAL-4.8-Ic-81/doc/Mesh_3/examples.txt:18: Warning: included file Mesh_3/mesh_polyhedral_implicit_function.cpp is not found. Check your EXAMPLE_PATH
    /tmp/CGAL-4.8-Ic-81/doc/Mesh_3/examples.txt:19: Warning: included file Mesh_3/mesh_polyhedral_surface_tolerance_region.cpp is not found. Check your EXAMPLE_PATH

**Comment by sloriot** [comment]
I fixed the PSP issue in 81d6383

**Comment by sloriot** [comment]
Those errors are not related by this PR. They are related to examples in dont_submit. I'll update the `examples.txt` files and merge the branch

File: Documentation/doc/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    )
+    add_dependencies(Documentation_test doc)
+
+    add_custom_target(Documentation_test_publish
+      ${PYTHON_EXECUTABLE} ${CGAL_DOC_TESTSUITE_SCRIPT} --output-dir ${CGAL_DOC_OUTPUT_DIR} --doc-log-dir ${CGAL_DOC_LOG_DIR} --publish ${CGAL_DOC_PUBLISH_DIR} --do-copy-results
+    )
+    add_dependencies(Documentation_test_publish doc)
+
+    set(CGAL_DOC_VERSION_H "/tmp/version.h" CACHE PATH "Path to CGAL/version.h")
+
```
Comment (lrineau) [inline Documentation/doc/CMakeLists.txt]: I think that *ugly* hack should be removed, now.

Comment (lrineau) [inline Documentation/doc/CMakeLists.txt]: The version should be passed as a CMake variable, instead of using a file in `/tmp`.

**Comment by lrineau** [comment]
Is there a documentation that explains how to build the documentation, now?

**Comment by bo0ts** [comment]
> Is there a documentation that explains how to build the documentation, now?

Like it used to: `cmake -DBUILD_DOC=TRUE /path/to/cgal; make doc`

File: Documentation/doc/CMakeLists.txt (hunk 1 above)
Comment (bo0ts) [inline Documentation/doc/CMakeLists.txt]: I found out what that hack was actually for. I'll open a separate PR to get rid of it.

**Pull request merged**